### PR TITLE
feat: signing order

### PR DESCRIPTION
**Pull request by ephraimduncan**
https://github.com/user-attachments/assets/439ca3f0-150e-4685-858f-5f136bd7a5ac

## Summary by CodeRabbit

- **New Features**
	- Introduced a signing order feature for document signing, allowing users to specify the order in which signers will receive documents.
	- Enhanced document editing with the ability to manage signing order directly in the submission workflow.
	- Implemented drag-and-drop functionality for reordering signers, updating their signing orders accordingly.
	- Expanded localization strings to support the signing order feature in multiple languages.
	- Added comprehensive coding guidelines and best practices for TypeScript development.

- **Bug Fixes**
	- Improved handling of recipient signing order to ensure that signing actions respect the specified order.

- **Documentation**
	- Updated localization strings to reflect new features and improve user experience.

### CODE REVIEW

**Comment by ephraimduncan** [comment]
@coderabbitai review

File: packages/ui/primitives/document-flow/add-fields.tsx [hunk 1]
```diff
@@ -478,9 +478,23 @@ export const AddFieldsFormPartial = ({
 
   const recipientsByRoleToDisplay = useMemo(() => {
     // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
-    return (Object.entries(recipientsByRole) as [RecipientRole, Recipient[]][]).filter(
-      ([role]) => role !== RecipientRole.CC && role !== RecipientRole.VIEWER,
-    );
+    return (Object.entries(recipientsByRole) as [RecipientRole, Recipient[]][])
```
Comment (catalinpit) [inline packages/ui/primitives/document-flow/add-fields.tsx:481]: should we split/try to improve this code?

**Comment by Mythie** [comment]
This is now ready to go but we need to deal with the failing tests. I'm assuming it'll be a symptom of our updated sorting logic throwing an error somewhere

File: packages/lib/server-only/document/complete-document-with-token.ts [hunk 2]
```diff
@@ -174,17 +138,48 @@ export const completeDocumentWithToken = async ({
...
     where: {
       documentId: document.id,
       signingStatus: {
         not: SigningStatus.SIGNED,
       },
+      role: {
+        not: RecipientRole.CC,
+      },
     },
+    // Composite sort so our next recipient is always the one with the lowest signing order or id
```
Comment (ephraimduncan) [inline packages/lib/server-only/document/complete-document-with-token.ts:155]: 😍

File: packages/ui/primitives/document-flow/add-signers.tsx [hunk 3]
```diff
@@ -207,125 +364,262 @@ export const AddSignersFormPartial = ({
...
+              ]}
+            >
+              <Droppable droppableId="signers">
+                {(provided) => (
+                  <div
+                    {...provided.droppableProps}
+                    ref={provided.innerRef}
+                    className="flex w-full flex-col gap-y-2"
+                  >
+                    {signers.map((signer, index) => (
```
[the remark this answers is not in the document]
Comment (ephraimduncan) [reply packages/ui/primitives/document-flow/add-signers.tsx:422]: cc: @goutham-documenso

File: apps/web/src/app/(signing)/sign/[token]/form.tsx [hunk 4]
```diff
@@ -150,6 +157,7 @@ export const SigningForm = ({ document, recipient, fields, redirectUrl }: Signin
                     fields={fields}
                     fieldsValidated={fieldsValidated}
                     role={recipient.role}
+                    disabled={!isRecipientsTurn}
```
[the remark this answers is not in the document]
Comment (ephraimduncan) [reply apps/web/src/app/(signing)/sign/[token]/form.tsx:160]: Lucas fixed it in his commit.

CleanShot 2024-09-16 at 11 01 30@2x

**Review by Mythie — APPROVED** [review]

**Review by Mythie — APPROVED** [review]

**Pull request merged**